### ISSUE TITLE
Proposal: Replacing the PromQL Parser with a generated parser

### ISSUE THREAD

**Issue by slrtbtfs** [prometheus/prometheus#6256]
# Replacing the PromQL Parser with a Generated Parser

### Current State of the Parser

Currently, the PromQL query engine utilizes a handwritten, recursive-descent parser, that was written five years ago and hasn't seen large changes since then.
It seems that none of the active members of the Prometheus project has a deep understanding of the parser code and is willing to maintain it or review complex PRs on it (based on the discussion in <https://github.com/prometheus/prometheus/pull/6061>).

There also seem to be some hidden features and edge cases that nobody is aware of (see <https://github.com/prometheus/prometheus/pull/6198> and <https://github.com/prometheus/prometheus/issues/6065>).

The code seems to be far more complex than necessary and would require significant refactoring before continuing to develop new features on it.

### Using the Parser for the Upcoming Language Server 

The upcoming PromQL language Server (see [proposal](https://github.com/slrtbtfs/promql-lsp/blob/master/doc/proposals/2019_promql-language-server.md)) requires a parser in order to analyze PromQL queries.

For purposes of compatibility, maintainability, and code reuse, it is intended for the language server to use the same PromQL parse as Prometheus.
To enable that use case, some enhancements to the existing parser have been [proposed](https://github.com/slrtbtfs/promql-lsp/blob/master/doc/proposals/2019_promql-parser-improvements.md).
These mainly concern the _output_ of the parser and have already been partly implemented.
However, some of the features described there, such as error recovery, would be difficult to implement in the existing parser.

#### Producing a Deeper Syntax Tree

Consider the following PromQL query:

    metric{label=="value", foo~="ba."}[5m] offset 3h

The corresponding syntax tree produced by the current parser looks something like the following:

    |---- MatrixSelector: metric{label=="value", foo~="ba."}[5m] offset 3h

This syntax tree is sufficient if the only goal is evaluating PromQL queries.
In fact, the PromQL query engine expects this simplified format.

However, it is insufficient for more involved use cases, such as: auto-completion that works almost anywhere in the code, generating helpful error messages, showing function signatures and documentation while editing queries, etc.
Instead, a syntax tree structured as follows would be more appropriate:

    |---- OffsetModifier: metric{label=="value", foo~="ba."}[5m] offset 3h
    . . . |---- MatrixSelector: metric{label=="value", foo~="ba."}[5m]
    . . . . . . |---- VectorSelector: metric{label=="value", foo~="ba."}
    . . . . . . . . . |---- LabelMatcher: label=="value"
    . . . . . . . . . . . . |---- Label: label
    . . . . . . . . . . . . |---- Value: "value"
    . . . . . . . . . |---- LabelMatcher: foo~="ba."
    . . . . . . . . . . . . |---- Label: foo 
    . . . . . . . . . . . . |---- Value: "ba."

#### It is easier to maintain

The complex task of recognizing the PromQL syntax is performed by the generated code.
Humans only have to deal with a comparably readable formal grammar and the code that builds up a syntax tree from already well structured input.

#### It explicitly documents the formal grammar of PromQL

A formal grammar provides a human- and machine-readable definition of what is a correct PromQL query.
This would eliminate edge cases and unknown features such as the ones described above.

#### It is possible to handle incomplete expressions

Consider the following incomplete PromQL input from a user:

    some_metric{la

By adding syntax rules for incomplete expressions, the parser could figure out that it got an incomplete vector selector with a certain structure.
A language server could then use this information to offer completions for all labels that occur together with `some_metric` and start with `la`.

#### It is easier to provide helpful error messages

All that needs to be done in order to add new error messages, is to add a grammar rule that recognizes this specific error class and add some code that handles this error when building up the syntax tree.

## Proposal

This document proposes replacing the current PromQL parser with one that is generated by [goyacc](https://godoc.org/golang.org/x/tools/cmd/goyacc).
Goyacc was chosen because yacc-generated parsers have been successfully used for decades.
Further, of the few alternatives for generating go parsers, none seem to be on par with goyacc.

The new parser should able to output two syntax tree formats:

* a deeper tree, where almost every token is assigned a node, designed to be used by applications that require additional context, such as the language server; and

* a simplified tree that looks almost like the current one, intended to be used by the PromQL query engine and any other application that does not need the additional context.

The parser should able to parse incomplete expressions.

The changes proposed in a [previous proposal](https://github.com/slrtbtfs/promql-lsp/blob/master/doc/proposals/2019_promql-language-server.md) remain valid.
The changes to the syntax tree and lexer that have already been implemented can be still used with the new parser.
The remaining changes, i.e. incomplete expression parsing and having an error list, will be implemented along with the newly proposed parser.

### lexer

Yacc relies on receiving a stream of tokens from  a lexer.
The new parser will continue to use the logic of the existing lexer.
The lexer output will be modified to satisfy the interfaces required by goyacc.
To be consistent with the usual yacc conventions, token names will be capitalized.

### yacc grammar

The core of the parser is a formal grammar of PromQL, which is the input for the parser generator and provides a precise and explicit definition of the PromQL syntax.
An extra rule must be declared for every class of error and incomplete expressions that the parser should handle.

The grammar is able to recognize all valid PromQL queries as well as common cases of incomplete or otherwise incorrect queries.

As an example, the part of the grammar recognizing an instant vector selector will look roughly like:

    selector:
        IDENTIFIER                          {$$ = newVectorSelector(nil, $1}
      | IDENTIFIER LBRACE          RBRACE   {$$ = newVectorSelector(nil, $1, $2, nil, $3)}
      | IDENTIFIER LBRACE matchers RBRACE   {$$ = newVectorSelector(nil, nil, $1, $2, $3}
      |            LBRACE          RBRACE   {$$ = newVectorSelector(nil, nil, $1, nil, $2)}
      |            LBRACE matchers RBRACE   {$$ = newVectorSelector(nil, $1, $2, $3,  $4)}
      | IDENTIFIER LBRACE matchers anything {$$ = newVectorSelector([]ErrCodes{
                                                                            ErrNoClosingBrace,
                                                                        }, $1, $2, $3,  $4)}
      |            LBRACE matchers anything {$$ = newVectorSelector([]ErrCodes{
                                                                            ErrNoClosingBrace,
                                                                        }, nil, $1, $2, $3)}

    matchers:
        matchers COMMA matcher              {$$ = newMatchers(nil, $1, $2, $3)}
      | matcher                             {$$ = newMatchers(nil, $1)}

    matcher:
        IDENTIFIER matchOp STRING           {$$ = newMatcher(nil, $1, $2, $3)}
      | IDENTIFIER matchOp anything         {$$ = newMatcher([]ErrCodes{
                                                                        ErrWrongType,
                                                                }, $1, $2, $3)}
      | IDENTIFIER anything                 {$$ = newMatcher([]ErrCodes{
                                                                        ErrNoMatchOp,
                                                                }, $1, $2)}

    matchOp:
        EQ                                  {$$ = $1}
      | NEQ                                 {$$ = $1}
      | REGEQ                               {$$ = $1}
      | REGNEQ                              {$$ = $1}

    
    anything:
        // All valid PromQL Syntax Elements
        ...
      | EOF                                 {$$ = $1}
      | error                               {$$ = $1}

### Building up the Abstract Syntax Tree (AST)

Every match in the grammar calls a function of the following type or simply returns the token if no special action is required:

    func ([]ErrCodes, ...*Expr) *Expr

Tokens implement the Expr interface, so no special distinction has to be made here.

These functions gradually build up a syntax tree and perform all error and type checking along the way, using the hints provided by the given `ErrCodes`

### The Expr Interface

All Syntax tree nodes must satisfy the `Expr` interface and implement at least the following functions:

    type Expr interface {
        // A string representation
        String()     string
        // Position
        Pos()        token.Pos
        EndPos()     token.Pos
        // Errors in the respective subtree
        Errors()     []*ParseErr
        // The Type of the Expr
        Type()       ExprType
        // Simplifies the subtree to allow fast Evaluation by the PromQL engine
        Simplify()   *Expr
        // Reports if the subtree is already simplified. If yes, the subtree is 
        // skipped by Simplify() for performance reasons
        Simplified() bool
    }

### AST Simplifying and Backwards Compatibility

The syntax tree produced by the new parser will provide an accurate and detailed representation of the query.
However, when evaluating queries in the PromQL query engine, it is better to have a less complex syntax tree.
In fact, the PromQL query engine expects matrix and vector selectors to be represented as a single node rather and as a complex tree.

In order to avoid completely breaking the query engine and any other existing users of the current PromQL parser, the `Expr` interface will have a function `Simplify()` that recurses through the subtree, tries to reduce the depth of the tree and changes matrix and vector selector to the format that the query engine currently expects. 

To produce this simplified, backwards-compatible syntax tree format, it suffices to call `Simplify()` on the root node.

It is possible to implement additional optimizations like pre-evaluation of scalar expressions in the `Simplify()` method.

### Scope of the change

The parser will be completely rewritten.

The output format and token naming conventions in the lexer will be changed to fit the conventions and interfaces of yacc.
Most of the actual logic will remain unchanged.

The changes to the query engine will be minimized by the AST Simplification mechanism.
The changes will be confined to some name changes in types and interfaces.

### Possible Regressions

By Rice's theorem, it may be impossible be certain whether the new parser accepts exactly the same syntax as the old one.

Less academically, it is possible that there exist edge cases in the current parser that the new parser does not handle because no one knows about them.
It is likely impossible to completely mitigate the risk of accidentally introducing a breaking incompatibility of this type.
This is a consequence of the fact that no formal definition of PromQL exists yet.

In order to reduce this risk, a large sampling of queries that work with the current parser should be collected and used to test the new parser.

---
Source of truth for this proposal: <https://github.com/slrtbtfs/promql-lsp/blob/master/doc/proposals/2019_generated%20promql%20parser.md>

---

**TL;DR:** This document proposes rewriting the current parser and replacing it with a generated one. This should make it more maintainable and easier to implement some features needed for the upcoming PromQL language server.

**Comment by slrtbtfs** [comment]
cc @juliusv @brian-brazil @brancz @squat @krasi-georgiev @simonpasquier @beorn7

**Comment by beorn7** [comment]
This totally makes sense to me, but I'm not an expert about parsers in general, and I have never really worked closely with the Prometheus parser in particular. Thus, I think input from various other people would be very valuable here.

@juliusv wrote the original parser, which _was_ generated. @fabxc wrote the current hand-coded one, mostly (IIRC) because the generated one wasn't creating very good error messages. Apparently, it was a misconception on our side back then that generated parsers cannot create good error messages. @tomwilkie explained that to me. All three probably have some valuable context to share here, in particular about more subtle aspects that @slrtbtfs might have missed, despite his thorough analysis, but that might turn out to be dangerous pitfalls later.

**Comment by juliusv** [comment]
I like it for all the reasons stated.

Couple of comments:

- Having two representation types (expanded and simplified) of the same PromQL expression adds extra complexity and concepts one needs to wrap their head around. How invasive would the changes need to be in the query engine to also use the expanded format, or does that immediately seem like a bad idea?
- Do `Simplify()` and `Simplified()` need to be defined on the `Expr` interface, or could they be external functions that work on all expected expression types, like we currently do a lot in the engine code (e.g. https://github.com/prometheus/prometheus/blob/74726367cf7a7e8d0332238defd2e7f4169030bd/promql/engine.go#L924). Especially if you're also thinking of doing arbitrary optimizations, it could be better to do this externally outside of the type in a way that is swappable.

**Comment by slrtbtfs** [comment]
The main reason for adding the `Simplify()` interface was to be able to more or less swap out the two parsers and avoid breaking things outside the parser. This keeps the proposal more constrained. 

Modifications in the engine required for using the new kind of syntax tree in the engine would be:
* Handling the cases created by new node types
* Adding a way to figure out the time and label information of complex, nested matrix- and vector selector subtrees. Ways to do that include, but are not limited to:
  * Creating helper functions for that task
  * Already carrying the information in the nodes and passing it down the tree when generating the syntax tree.
* Also, we'd possibly have to change other parts inside or outside prometheus that are using the parser, e.g. [these](https://github.com/prometheus/prometheus/search?q=promql.ParseMetricSelector&unscoped_q=promql.ParseMetricSelector)

It would be possible for `Simplify()` and `Simplified()` to be exported as functions instead of methods. Actually it isn't even necessary to export them at all. Instead we could just export two different `Parse()` functions which produce different syntax tree formats. 

The main reason for using these `switch expr.(type)` constructs is [performance](https://github.com/golang/go/issues/20116), right? It should indeed be possible to replace any interface calls with these constructs, at the cost of reduced readability.

**Comment by juliusv** [comment]
> The main reason for using these switch expr.(type) constructs is performance, right?

No, I don't think performance matters at all here. IIRC @fabxc just preferred having code acting on AST nodes externally instead of having all the logic baked into the AST node types themselves.

**Comment by juliusv** [comment]
> Adding a way to figure out the time and label information of complex, nested matrix- and vector selector subtrees.

Yeah, that seems like it would introduce the most verbosity / extra complexity, I think the rest sounds kind of ok. Just wanted to indicate that from my side changing the engine would be ok too, rather than having two AST representations for the same thing. But will leave the tradeoff judgement to you.

**Comment by juliusv** [comment]
> The main reason for adding the `Simplify()` interface was to be able to more or less swap out the two parsers and avoid breaking things outside the parser.

You can still swap out the parser as long as the AST nodes stay the same (would be good to keep them relatively stable for outside use in general)...

**Comment by slrtbtfs** [comment]
> You can still swap out the parser as long as the AST nodes stay the same (would be good to keep them relatively stable for outside use in general)...

Or at least keeping around the existing AST Fields. Probably it is still needed to add new node types, however.

**Comment by slrtbtfs** [comment]
> Just wanted to indicate that from my side changing the engine would be ok too, rather than having two AST representations for the same thing. But will leave the tradeoff judgement to you.

If everybody is ok with that, I'd prefer making changes to the engine over having two syntax tree formats.

**Comment by juliusv** [comment]
@slrtbtfs Totally. Having two implementations of the same thing always sucks for everyone, so unless you think the engine changes would become totally crazy, I would prefer the engine to be changed.

**Comment by slrtbtfs** [comment]
> > The main reason for using these switch expr.(type) constructs is performance, right?
> 
> No, I don't think performance matters at all here. IIRC @fabxc just preferred having code acting on AST nodes externally instead of having all the logic baked into the AST node types themselves.

For cases they don't create performance problems, I have a strong preference for interfaces instead of `switch expr.(type)` constructs:
* Interfaces avoid code duplication and improve readability (see https://github.com/prometheus/prometheus/pull/6205/commits/27d58006510b3b930e2df8645f82df6b6502ea8f and https://github.com/prometheus/prometheus/pull/6205/commits/abec09e9a144a02f1c386bbff60a7fa91710e138)
* For interfaces the static analysis baked into the compiler makes sure that no cases are forgotten, so there is no need for the `default: panic("We forgot to implement this")` case. This means that most of the problems caused by changes in the AST are already caught at compile time. 
* Switches tend to generate huge, highly complex and hard to read functions that can span hundreds of lines.

**Comment by slrtbtfs** [comment]
> @slrtbtfs Totally. Having two implementations of the same thing always sucks for everyone, so unless you think the engine changes would become totally crazy, I would prefer the engine to be changed.

They should be less work than the `Simplify()` logic.

**Comment by brian-brazil** [comment]
Our PromQL benchmarks include the parsing, so if it matters it'll show up there.

**Comment by juliusv** [comment]
> For cases they don't create performance problems, I have a strong preference for interfaces instead of switch expr.(type) constructs:

Yeah, makes sense. That is actually what my initial engine + AST looked like before the rewrite by @fabxc, so I guess he'd be in the best position to provide more historical context on why it was changed to be switch-y instead. But seems like in this case we're avoiding the question altogether by just changing the engine then instead of adding `Simplify()` / `Simplified()` methods.

**Comment by juliusv** [comment]
> Interfaces avoid code duplication and improve readability (see 27d5800 and abec09e)

Take a look at this from 2015 :) https://github.com/prometheus/prometheus/blob/0.10.0/rules/ast/walk.go

**Comment by slrtbtfs** [comment]
I see, so all I do is bringing PromQL back to where it was in 2015 :)

**Comment by slrtbtfs** [comment]
Some updates on this:

## Lexer

After having done some work on the lexer I found out, that large parts of PromQL error handling happen in the lexer. So if we want to get decent error messages, it might make sense to replace the lexer, too.

However this can happen mostly independently to a parser rewrite, so this proposal isn't really affected.

## AST Simplifying

After the feedback in this thread, I'd just ditch  this part of the proposal and make some small changes to the engine where necessary.

## Performance

After the performance improvements in #6356, the performance of the parser has become more fragile, i.e. even tiny changes to the parser have a visible performance impact in benchmarks.

Given that this proposal seeks to increase the size of the AST, i.e. do more allocations, I don't expect performance to get better.

Also, when only parsing a single query, setting up position handling introduces a constant penalty of about 1000 ns per query (which can be avoided using some tricks if desired).

However, I'm quite confident that performance at the end won't be worse than it was before  #6356.

For Prometheus itself this doesn't really matter anyway, since in the PromQL engine benchmarks less than 10% of query evaluation time is spent in the parser for small range queries. For larger range queries the parser performance becomes irrelevant.

## How to replace one parser with another

Currently there is still this huge PR #6061 lying around. A lot of what is done there would be far simpler to implement with a generated parser, so I'd suggest to just ditch this PR and include that functionality later and in smaller chunks.
 
For the parser replacement I'd suggest the following procedure, split into multiple phases.

### Preparing the lexer
Rename and refactor a few things in the lexer to make it compatible with the interfaces and conventions of the parser generator. This might also include some behavior changes around dealing with white space and comments.

### Gradual replacement
The current parser is built in a recursive way. This can be seen as one main parser spawning some child parsers. When viewing this as a tree of parsers there exist some leaf parsers, e.g. the `labelMatchers` parser, which can be replaced by a generated parser without too much breakage. 

Replacing those leaf parsers with generated ones should be possible in relatively small, isolated PRs while still keeping the parser in a working state. From there, it is possible to gradually work up the tree.

No breaking changes to the AST shall be made during this phase.

### AST improvements

Once the parser is fully generated, changes to the AST should be relatively easy to implement. These include porting the changes of #6061 to the new generated parser and doing the AST changes described in the proposal.

### Improving error handling

This consist of two things:
* Getting rid of the current panic-recover error handling and instead making errors part of the syntax tree.
* Do type checking at AST generation time. Currently it happens after parsing is completed in a recursive traversion of the syntax tree.

## Summary 
This is a suggestion on how to go on with the parser replacement. If feedback on this is positive I'm gonna start working on these changes soon. Also it would be nice to know if there is enough maintainer capacity to review and merge these upcoming changes.

**Comment by juliusv** [comment]
Generally it all sounds good, provided error messages stay equally good or get better.

> Also it would be nice to know if there is enough maintainer capacity to review and merge these upcoming changes.

How does the reviewer capacity situation look at e.g. Red Hat? I might try to take a look as well here and there, but can't guarantee as much as people who can do it as part of a paid job :)

**Comment by simonpasquier** [comment]
I'll add this to my review list for next week (though I'm not that knowledgeable about the parser code). @brancz is probably going to have a look to but I think he'll be trying to recover from KubeCon :smile:

**Comment by slrtbtfs** [comment]
> Generally it all sounds good, provided error messages stay equally good or get better.
Better error messages are one of the main goals of this proposal.

**Comment by juliusv** [comment]
Yup, that's what I assumed :) So sounds good.

**Comment by slrtbtfs** [comment]
The worst error messages we currently have ("could not parse remaining input ...")  are created in the lexer and there isn't really a way to improve them without massive lexer changes.

So there might be another lexer replacement proposal after the parser replacement is done. However replacing the lexer should be much easier than replacing the parser, if the new parser is designed with the possibility of such changes in mind.

**Comment by slrtbtfs** [comment]
> The current parser is built in a recursive way. This can be seen as one main parser spawning some child parsers. When viewing this as a tree of parsers there exist some leaf parsers, e.g. the labelMatchers parser, which can be replaced by a generated parser without too much breakage.

> Replacing those leaf parsers with generated ones should be possible in relatively small, isolated PRs while still keeping the parser in a working state. From there, it is possible to gradually work up the tree.

I've actually visualized this "Tree" now and it turns out there are a lot of cycles. To remove these, it will be necessary to do some larger PRs replacing larger cyclic subgraphs towards the end of the rewrite.

So, this is what the PromQL parser currently looks like:

![PromQL parser structure](https://user-images.githubusercontent.com/15368509/70924866-90e71500-202a-11ea-9690-a55609b28bd5.png)

**Comment by brian-brazil** [comment]
Why do you think you can remove cycles in an LL(1) grammar?

**Comment by slrtbtfs** [comment]
I don't think that.

The reason why cycles are problematic is, that when replacing the parser only subgraphs with no outgoing edges can be replaced by the generated parser.

This makes it possible to replace leaves in small distinct PRs. That's what the PRs I've written so far did.

If there are cycles, I have to replace a larger subgraph in one PR, which will cause larger PRs that are harder to write and review.

**Comment by slrtbtfs** [comment]
Does the above make sense?

**Comment by brian-brazil** [comment]
That makes sense.

**Comment by slrtbtfs** [comment]
While there are still a lot of upcoming improvements for the parser, the actual replacement work is done now, so I guess I can close this.

**Comment by juliusv** [comment]
Nice!

**Comment by slrtbtfs** [comment]
> The worst error messages we currently have ("could not parse remaining input ...") are created in the lexer and there isn't really a way to improve them without massive lexer changes.

Worth pointing out that we actually completely got rid of this particularly bad error message as a side effect.
